### PR TITLE
Live chatv2 giga super mega merge

### PR DESCRIPTION
**Pull request by Pommfritz0308**
**Backend Changes:**

1. ASGI Configuration (asgi.py):
   - Added WebSocket routes for the live_chat app, ensuring chat WebSockets are handled correctly by adding live_chat.routing.websocket_urlpatterns.

2. Settings (settings.py):
   - Added live_chat to INSTALLED_APPS for managing chat functionality.

3. URL Routing (urls.py):
   - Added a new API route for chat: /api/chat/, connecting chat views and WebSocket routing to the application.

4. Live Chat App:
   - Models (models.py):
     - Message: Handles chat messages with fields for sender, receiver, timestamp, and content.
     - Relationship: Manages friend requests and chat relationships between users.
   - Admin (admin.py): Registered Message and Relationship models.
   - Views (views.py):
     - Added views to manage relationships between users (RelationshipStatusView).
     - API endpoints for retrieving and updating relationship status between two users.
   - WebSocket (consumers.py):
     - Implemented ChatConsumer to manage WebSocket connections, sending and receiving messages, and updating the frontend dynamically.
   - Routing (routing.py): Defined WebSocket URL patterns for live chat.

5. Django Admin:
   - Registered Message and Relationship models for Django admin interface.

6. New User Management API (views.py in usermanagement):
   - Added endpoint /users/<str:displayname>/ to retrieve a user’s ID by their display name, used for user search input.

**Frontend Changes:**

1. New Live Chat Interface (live_chat.html):
   - HTML Structure: A responsive chat interface with a chat window, user search field, user list, and message display area.
   - CSS Styling: Defined styles for chat messages, including separate styles for sent (.message-sender) and received (.message-receiver) messages, and timestamp styling.

2. Frontend JavaScript (live-chat.js and chatHandler.js):
   - WebSocket Initialization: Dynamically sets up a WebSocket connection when the chat page is loaded.
   - Real-Time Updates: Handles real-time message sending, receiving, and updating the chat interface with the latest info.
   - User Management: Includes a dynamic user list and allows users to send friend requests, accept, or block requests in real-time + manage game invitations with befriended users + user blocking mechanism to hide your online status from blocked user's active user list and block all of their messages
   - Chat and game notifications across all pages (except unregistered paths and /game)

3. Chat Room (chat_room.html):
   - Introduced a separate chat room page with message history, input handling, and dynamic user interaction through WebSockets.

4. Frontend Routing (app.js and router.js):
   - Added routes for the new chat and chat room pages.
   - Integrated chat-specific logic into the existing frontend router, ensuring chat functionality is properly initialized.

5. CSS Changes:
   - Chat-Specific Styles (chat.css): Added styles for the chat interface, including the chat window, message items, and user list layout.

### CODE REVIEW

File: backend/pong/config.toml
Comment (stevebalk) [inline backend/pong/config.toml:1]: 10 seconds is a bit long, isn't it?

File: backend/pong/consumers/matchmaking_consumer.py
Comment (stevebalk) [inline backend/pong/consumers/matchmaking_consumer.py:1]: Please change it to a log message or remove it if it's no longer needed.

Comment (stevebalk) [inline backend/pong/consumers/matchmaking_consumer.py:1]: Please change it to a log message or remove it if it's no longer needed.

File: backend/pong/data_managment/user.py
Comment (stevebalk) [inline backend/pong/data_managment/user.py:1]: Please remove the methods if we don't need them (You don't use them)

File: backend/pong/match/match_session_handler.py
Comment (stevebalk) [inline backend/pong/match/match_session_handler.py:1]: Please change it to a log message or remove it if it's no longer needed.

File: backend/live_chat/consumers.py [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+                'requester_profile_picture': requester_profile_picture_url
+            })
+
+        await self.send_message_to_user(user_id, {
+            'message': pending_requests_info,
+            'message_type': 'pending_requests',
+            'message_key': 'requests'
+        })
+    
+    # async def send_current_game_registration(self, user_id):
```
Comment (memaar92) [inline backend/live_chat/consumers.py]: remove if no longer needed. Also applies to other commented code

File: backend/pong/config.toml
Comment (memaar92) [reply backend/pong/config.toml:1]: Yeah, could be a bit shorter. Yesterday, I thought 5secs is too short, but now seeing it in action for the first time, I agree that 10secs might be a bit long

File: frontend/js/chatHandler.js [hunk 2]
```diff
@@ -0,0 +1,1228 @@
...
+      console.error('Failed to reconnect:', error);
+    }
+  }
+  
+  async onClose(event, context) {
+    console.log('WebSocket connection closed:', event.code, event.reason);
+    
+    if (event.code === 1006) {
+      try {
+        await this.checkToken();
```
Comment (memaar92) [inline frontend/js/chatHandler.js]: Why is this exactly needed? What would be the issue if the token is not checked?

Comment (memaar92) [reply frontend/js/chatHandler.js]: Also, what is event code 1006?

File: backend/live_chat/templates/live_chat.html [hunk 3]
```diff
@@ -0,0 +1,304 @@
+<!DOCTYPE html>
```
Comment (memaar92) [inline backend/live_chat/templates/live_chat.html]: Wird die noch genutzt?

File: backend/live_chat/views.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+from .models import Relationship
+from django.core.serializers import serialize
+from django.contrib.auth import get_user_model
+import json
+from channels.layers import get_channel_layer
+from asgiref.sync import async_to_sync
+
+User = get_user_model()
+
+def live_chat(request):
```
Comment (memaar92) [inline backend/live_chat/views.py]: noch benutzt?

Comment (Pommfritz0308) [reply backend/live_chat/views.py]: nein

File: frontend/js/chatHandler.js [hunk 5]
```diff
@@ -31,7 +31,7 @@ class ChatHandler {
 
     console.log('Creating new WebSocket connection');
     console.log('Checking token');
-    await this.checkToken();
+    await this.refreshToken();
```
Comment (memaar92) [inline frontend/js/chatHandler.js]: Immer den Token zu refreshen, egal, ob ein gültiger Access token existiert scheint problematisch zu sein.
Bsp.: Ich logge mich ein und bleibe auf dem main-menu bis der access token abläuft. Dann klicke ich auf Tournament. Sowohl der ChatHandler als auch Wayne's Frontend schicken dann quasi zeitgleich refresh Token-Anfragen. Der eine scheint aber nicht mehr valide zu sein und ich werde ausgeloggt

**Review by memaar92 — CHANGES_REQUESTED** [review]
siehe Kommentar zum Token-Refresh. Bitte nochmal ausgiebig testen vor nächstem Push

**Comment by wrottger** [comment]
#89 should be fixed now

**Review by memaar92 — APPROVED** [review]
Major issues have been fixed and main chat features are working. Well done

**Pull request merged**